### ISSUE TITLE
Make build output idempotent and use shorter hash

### ISSUE THREAD

**Issue by jakub-g** [ariatemplates/ariatemplates#1042]
There are two things with our build / packing that are wrong / suboptimal to me. They both concern appending of hashes for versioning.

TL;DR: look to **"Actions"** in bold

Let's state the facts first.
- We're using 128 bit MD5 for the hashes
- Why we're using the hashes at all
  1. So that the application can safely swap the version of AT they use; after the bundle contents changed, the URLs of deps of the bootstrap will have new hashes
     and in case the users had versions in their cache, the browser will ignore it and will make requests for the new ones instead of reading stale cache data
  2. So that the applications can deploy multiple version of AT to the same folder
  3. In case though that the contents of the bundle didn't change, there's no need to to request for it - the hash guarantees we still have the good version in cache
# Non-idempotent builds

First thing is, during each build, _for each bundle_, we append the header at the beginning of it, and it _always_ contains the version number and build date. Even for patches.

That means that if you release say 1.4.17A that changes **one line in one file** with respect to 1.4.17, it'll have `Aria Templates 1.4.17A - 20 Mar 2014` in the banners for each bundle.
That effectively means there's _absolutely no gain_ in having various builds of AT in the same directory, because each file on each version will be hashed to a different hash.

Also, if you build 1.x.x one day and rebuild from the same source tomorrow, all the hashes are changing.

To sum up, iii) doesn't hold anymore - the _real_ content of our files didn't change yet we're requesting new ones.

All above applies to non-patch versions too, though usually most of the bundles change at least a bit between the releases (but not of them).

**Action:** IMO we could just add date & version number to `ariatemplates.js` and in others, just have a generic banner without timestamp / version number.
# MD5 is an overkill

We append MD5 to stamp the release but I think it's an overkill given what we want to achieve.
MD5 produces 128 bit output which is 32 bytes string using [0-f] alphabet. All of those are part of the urlmap which is appended to the bootstrap file, and those hashes are significant part of the urlmap.
1. 128 bit is way too much than needed
2. Alphabet 0-f is suboptimal

What we can do is to use "md5value".toString(36) which recodes the hash to [0-z] alphabet and solely this reduced the length from 32 bytes to 25 bytes per each hash string.
Other thing, we can halve the MD5 value (though some people say it's an antipattern) to have less bits. Still should be good enough for our case.
Even better, go with 32-bit murmur hash [1][2] and produce the hash with .toString(36) (outputs sth like 'ejn3zz') Murmur hash is apparently blazing fast [4] and very well distributed.

32 bits of output from murmur means ~4 bln of possible output hashes.
Let's say we have 100 versions of AT deployed on the server in the same folder (they're doing it really really wrong, but never underestimate what can happen!) and each version of the particular bundle is different by 1 byte at least.

Assuming the hash is as good as they claim, having a collision with 100 files with 4 billion possible hashes is 10^-6 order of magnitude. In practice there'll be less versions and the probability will be near zero.

I've done some checkings, taking just the **urlmap** from AT.com page:

```
original: 16,719 (md5) -> 13,519 (half md5) -> 11,719 (murmur base36)
gzipped:   3,813 (md5) -> 2,889  (half md5) -> 2,317  (murmur base36)
```

So we can save 5kB plaintext and 1.5kB gzipped switching to shorter hash. Also, there'll be a bit less traffic to digest on servers due to shorter URLs.
That doesn't sound like much, but small things add up [3] 

**Action:** Change MD5 (128 bit, 32 bytes string) to murmur-v3/base36 (32 bits, 7 bytes string) or at least MD5-base36encoded (still 128 bit, but 25 bytes string)

[1] http://en.wikipedia.org/wiki/MurmurHash
[2] https://www.npmjs.org/package/murmurhash-js
[3] http://chrishateswriting.com/post/68794699432/small-things-add-up
[4] http://jsperf.com/murmurhash-3-vs-md5/3
